tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,559 @@
+		if err := tb.sendProposalVote(epoch); err != nil {
+			tb.Log.With().Error("Failed to send first voting message",
+				log.Uint64("epoch_id", uint64(epoch)),
+				log.Err(err))
+		}
+	}()
+
+	go func() {
+		if err := tb.weakCoin.PublishProposal(epoch, firstRound); err != nil {
+			tb.Log.With().Error("Failed to publish weak coin proposal",
+				log.Uint64("epoch_id", uint64(epoch)),
+				log.Uint64("round", uint64(firstRound)),
+				log.Err(err))
+		}
+	}()
+
+	// For K rounds: In each round that lasts δ, wait for proposals to come in.
+	// For next rounds,
+	// wait for δ time, and construct a message that points to all messages from previous round received by δ.
+	for round := firstRound + 1; round <= tb.lastRound(); round++ {
+		select {
+		case <-ticker.C:
+			go func(epoch types.EpochID, round types.RoundID) {
+				if err := tb.sendVotesDelta(epoch, round); err != nil {
+					tb.Log.With().Error("Failed to send voting messages",
+						log.Uint64("epoch_id", uint64(epoch)),
+						log.Uint64("round", uint64(round)),
+						log.Err(err))
+				}
+			}(epoch, round)
+
+			go func(epoch types.EpochID, round types.RoundID) {
+				if err := tb.weakCoin.PublishProposal(epoch, round); err != nil {
+					tb.Log.With().Error("Failed to publish weak coin proposal",
+						log.Uint64("epoch_id", uint64(epoch)),
+						log.Uint64("round", uint64(round)),
+						log.Err(err))
+				}
+			}(epoch, round)
+
+		case <-tb.CloseChannel():
+			return nil
+		}
+	}
+
+	tb.waitAfterLastRoundStarted()
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) runProposalPhase(epoch types.EpochID) error {
+	// take all ATXs received in last epoch (i - 1)
+	atxList := ATXIDList(tb.epochATXGetter.GetEpochAtxs(epoch - 1))
+
+	// concat them into a single proposal message
+	m := NewProposalMessage(epoch, atxList)
+
+	serializedMessage, err := types.InterfaceToBytes(m)
+	if err != nil {
+		return fmt.Errorf("serialize proposal message: %w", err)
+	}
+
+	tb.Log.With().Debug("Serialized proposal message",
+		log.String("message", string(serializedMessage)))
+
+	if err := tb.net.Broadcast(TBProposalProtocol, serializedMessage); err != nil {
+		return fmt.Errorf("broadcast proposal message: %w", err)
+	}
+
+	proposalsHash := atxList.Hash()
+
+	tb.timelyProposalsMu.Lock()
+
+	if _, ok := tb.timelyProposals[epoch]; !ok {
+		tb.timelyProposals[epoch] = make(map[types.Hash32]struct{})
+	}
+
+	tb.timelyProposals[epoch][proposalsHash] = struct{}{}
+
+	tb.timelyProposalsMu.Unlock()
+
+	if err := tb.weakCoin.PublishProposal(epoch, 1); err != nil {
+		tb.Log.With().Error("Failed to publish weak coin message",
+			log.Uint64("epoch_id", uint64(epoch)),
+			log.Err(err))
+	}
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) sendProposalVote(epoch types.EpochID) error {
+	tb.setCurrentRound(epoch, firstRound)
+	// round 1, send hashed proposal
+	// create a voting message that references all seen proposals within δ time frame and send it
+	votesFor, votesAgainst := tb.calcVotesFromProposals(epoch)
```

Comment:
generally - when creating a hash from a list we must sort it beforehand so that for same values we'll get the same hash
I don't see an ordering issue with your code but please double check